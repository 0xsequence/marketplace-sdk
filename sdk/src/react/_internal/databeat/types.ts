import type { MarketplaceKind, OrderbookKind } from '../api';

export enum EventType {
	BUY_ITEMS = 0,
	SELL_ITEMS = 1,
	CREATE_LISTING = 2,
	CREATE_OFFER = 3,
	CANCEL_LISTING = 4,
	CANCEL_OFFER = 5,
	TRANSACTION_FAILED = 6,
}

interface PropsEvent {
	[key: string]: string;
}

interface NumsEvent {
	[key: string]: number;
}

interface Transaction extends PropsEvent {
	chainId: number;
	txnHash: string;
}

interface TradeItemsInfo extends PropsEvent {
	marketplaceKind: MarketplaceKind;
	collectionAddress: string;
	currencyAddress: string;
	currencySymbol: string;
}

interface TradeItemsValues extends NumsEvent {
	currencyValueDecimal: number;
	currencyValueRaw: number;
}

export interface TrackBuyItems {
	props: TradeItemsInfo & Transaction;
	nums: TradeItemsValues;
}

export interface TrackSellItems {
	props: TradeItemsInfo & Transaction;
	nums: TradeItemsValues;
}

interface ListOfferItemsInfo extends PropsEvent {
	orderbookKind: OrderbookKind;
	collectionAddress: string;
	currencyAddress: string;
	currencySymbol: string;
}

interface ListOfferItemsValues extends NumsEvent {
	currencyValueDecimal: number;
	currencyValueRaw: number;
}

export interface TrackCreateListing {
	props: ListOfferItemsInfo & Transaction;
	nums: ListOfferItemsValues;
}

export interface TrackCreateOffer {
	props: ListOfferItemsInfo & Transaction;
	nums: ListOfferItemsValues;
}

export interface TrackTransactionFailed extends Transaction, PropsEvent {}
