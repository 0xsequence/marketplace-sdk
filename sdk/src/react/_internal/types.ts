import type { Address } from 'viem';
import { z } from 'zod';
import type { ContractType, CreateReq } from '../../types';
import type { MarketplaceKind } from './api';

export const QueryArgSchema = z
	.object({
		enabled: z.boolean().optional(),
	})
	.optional();

export type QueryArg = z.infer<typeof QueryArgSchema>;

export const CollectableIdSchema = z.string().or(z.number());

export const AddressSchema = z.string().transform((val, ctx) => {
	const regex = /^0x[a-fA-F0-9]{40}$/;

	if (!regex.test(val)) {
		ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: `Invalid Address ${val}`,
		});
	}

	return val as Address;
});

export enum StoreType {
	MARKETPLACE = 'marketplace',
	SHOP = 'shop',
}

export type CollectionType = ContractType.ERC1155 | ContractType.ERC721;

type TransactionStep = {
	exist: boolean;
	isExecuting: boolean;
	execute: () => Promise<void>;
};

export type TransactionSteps = {
	approval: TransactionStep;
	transaction: TransactionStep;
};

export enum TransactionType {
	BUY = 'BUY',
	SELL = 'SELL',
	LISTING = 'LISTING',
	OFFER = 'OFFER',
	TRANSFER = 'TRANSFER',
	CANCEL = 'CANCEL',
}

export interface BuyInput {
	orderId: string;
	collectableDecimals: number;
	marketplace: MarketplaceKind;
	quantity: string;
}

export interface SellInput {
	orderId: string;
	marketplace: MarketplaceKind;
	quantity?: string;
}

export interface ListingInput {
	contractType: ContractType;
	listing: CreateReq;
}

export interface OfferInput {
	contractType: ContractType;
	offer: CreateReq;
}

export interface CancelInput {
	orderId: string;
	marketplace: MarketplaceKind;
}
