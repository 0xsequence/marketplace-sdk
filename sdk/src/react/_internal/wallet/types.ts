export interface WalletConnectionHandler {
  openConnectModal: () => void;
  isConnected: boolean;
  address?: string;
  getAddress?: () => Promise<string>;
}