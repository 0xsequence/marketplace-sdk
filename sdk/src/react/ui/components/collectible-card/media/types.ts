import type { ReactNode } from 'react';

type ContentType = 'image' | 'video' | 'html' | '3d-model' | null;

type ContentTypeState = {
	type: ContentType;
	loading: boolean;
	failed: boolean;
};

type MediaProps = {
	name?: string;
	assets: (string | undefined)[];
	assetSrcPrefixUrl?: string;
	className?: string;
	supply?: number;
	isLoading?: boolean;
	fallbackContent?: ReactNode;
};

export type { ContentType, ContentTypeState, MediaProps };
