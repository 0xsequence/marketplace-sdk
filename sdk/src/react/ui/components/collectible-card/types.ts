export enum CollectibleCardType {
	SHOP = 'shop',
	MARKETPLACE = 'marketplace',
	INVENTORY = 'inventory',
}
