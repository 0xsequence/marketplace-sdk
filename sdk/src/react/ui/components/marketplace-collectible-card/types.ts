import type { TokenMetadata as MetadataTokenMetadataType } from '@0xsequence/metadata';
import type { Address } from 'viem';
import type { CollectibleCardAction, MarketplaceType } from '../../../../types';
import type {
	CollectibleOrder,
	ContractType,
	TokenMetadata as MarketplaceTokenMetadataType,
	Order,
	OrderbookKind,
} from '../../../_internal';

// Base properties shared by all collectible card types
type MarketplaceCardBaseProps = {
	collectibleId: string;
	chainId: number;
	collectionAddress: Address;
	collectionType?: ContractType;
	assetSrcPrefixUrl?: string;
	cardLoading: boolean;
	marketplaceType?: MarketplaceType;
	isTradable?: boolean;
};

// Properties specific to Shop card
type ShopCardSpecificProps = {
	salesContractAddress: Address;
	tokenMetadata: MarketplaceTokenMetadataType | MetadataTokenMetadataType;
	salePrice:
		| {
				amount: string;
				currencyAddress: Address;
		  }
		| undefined;
	saleStartsAt: string | undefined;
	saleEndsAt: string | undefined;
	quantityDecimals: number | undefined;
	quantityInitial: string | undefined;
	quantityRemaining: string | undefined;
	unlimitedSupply?: boolean; // it's useful for 1155 tokens
};

// Properties specific to marketplace and inventory cards
type MarketCardSpecificProps = {
	orderbookKind?: OrderbookKind;
	collectible: CollectibleOrder | undefined;
	onCollectibleClick?: (tokenId: string) => void;
	onOfferClick?: ({
		order,
		e,
	}: {
		order?: Order;
		e: React.MouseEvent<HTMLButtonElement>;
	}) => void;
	balance?: string;
	balanceIsLoading: boolean;
	/**
	 * Callback function that is called when the user attempts to perform an action
	 * (such as buying or making an offer) that they are not permitted to do.
	 *
	 * This function is invoked in the following scenario:
	 *
	 * - When a disconnected user clicks on "Buy Now" and is prompted to connect
	 *   their wallet. After connecting, if it is determined that the user is
	 *   already the owner of the collectible, this callback is triggered to inform
	 *   them that they cannot perform the action (e.g., buying their own collectible).
	 *
	 * @param action - The action that the user cannot perform, which can be either
	 * CollectibleCardAction.BUY or CollectibleCardAction.OFFER.
	 */
	onCannotPerformAction?: (
		action: CollectibleCardAction.BUY | CollectibleCardAction.OFFER,
	) => void;
	prioritizeOwnerActions?: boolean;
};

// Complete CollectibleCardProps with all possible properties and card type
type MarketplaceCollectibleCardProps = MarketplaceCardBaseProps &
	Partial<MarketCardSpecificProps & ShopCardSpecificProps>;

type ShopCollectibleCardProps = MarketplaceCardBaseProps &
	ShopCardSpecificProps & {
		marketplaceType: 'shop';
	};
type MarketCollectibleCardProps = MarketplaceCardBaseProps &
	MarketCardSpecificProps & {
		marketplaceType?: 'market';
	};

type CollectibleCardProps =
	| ShopCollectibleCardProps
	| MarketCollectibleCardProps;

export type {
	MarketplaceCardBaseProps,
	ShopCardSpecificProps,
	MarketCardSpecificProps,
	MarketplaceCollectibleCardProps,
	ShopCollectibleCardProps,
	MarketCollectibleCardProps,
	CollectibleCardProps,
};
