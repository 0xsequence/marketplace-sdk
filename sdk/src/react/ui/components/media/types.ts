type ContentType = 'image' | 'video' | 'html' | '3d-model' | null;

type ContentTypeState = {
	type: ContentType;
	loading: boolean;
	failed: boolean;
};

type MediaProps = {
	name?: string;
	assets: (string | undefined)[];
	assetSrcPrefixUrl?: string;
	className?: string;
	isLoading?: boolean;
};

export type { ContentType, ContentTypeState, MediaProps };
