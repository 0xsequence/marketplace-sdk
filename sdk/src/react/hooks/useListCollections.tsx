'use client';

import { useQuery } from '@tanstack/react-query';
import type { Optional } from '../_internal';
import {
	type FetchListCollectionsParams,
	type ListCollectionsQueryOptions,
	listCollectionsQueryOptions,
} from '../queries/listCollections';
import { useConfig } from './useConfig';
import { useMarketplaceConfig } from './useMarketplaceConfig';

export type UseListCollectionsParams = Optional<
	ListCollectionsQueryOptions,
	'config' | 'marketplaceConfig'
>;

/**
 * Hook to fetch collections from marketplace configuration
 *
 * Retrieves all collections configured in the marketplace, with optional filtering
 * by marketplace type. Combines metadata from the metadata API with marketplace
 * configuration to provide complete collection information.
 *
 * @param params - Configuration parameters
 * @param params.marketplaceType - Optional filter by marketplace type
 * @param params.query - Optional React Query configuration
 *
 * @returns Query result containing array of collections with metadata
 *
 * @example
 * Basic usage:
 * ```typescript
 * const { data: collections, isLoading } = useListCollections();
 *
 * if (isLoading) return <div>Loading collections...</div>;
 *
 * return (
 *   <div>
 *     {collections?.map(collection => (
 *       <div key={collection.itemsAddress}>
 *         {collection.name}
 *       </div>
 *     ))}
 *   </div>
 * );
 * ```
 *
 * @example
 * Filtering by marketplace type:
 * ```typescript
 * const { data: marketCollections } = useListCollections({
 *   marketplaceType: 'market'
 * });
 * ```
 */
export function useListCollections(params: UseListCollectionsParams = {}) {
	const defaultConfig = useConfig();
	const { data: marketplaceConfig } = useMarketplaceConfig();

	const {
		config = defaultConfig,
		marketplaceConfig: paramMarketplaceConfig,
		...rest
	} = params;

	const queryOptions = listCollectionsQueryOptions({
		config,
		marketplaceConfig: paramMarketplaceConfig || marketplaceConfig,
		...rest,
	});

	return useQuery({
		...queryOptions,
	});
}

export { listCollectionsQueryOptions };

export type { FetchListCollectionsParams, ListCollectionsQueryOptions };
