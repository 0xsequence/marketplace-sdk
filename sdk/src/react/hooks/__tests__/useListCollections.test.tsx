import { renderHook, server, waitFor } from '@test';
import { http, HttpResponse } from 'msw';
import type { Address } from 'viem';
import { describe, expect, it } from 'vitest';
import { MarketplaceType, OrderbookKind } from '../../../types';
import {
	createLookupMarketplaceConfigHandler,
	mockConfig,
} from '../../_internal/api/__mocks__/builder.msw';
import {
	mockEthCollection,
	mockPolCollection,
} from '../../_internal/api/__mocks__/metadata.msw';
import { useListCollections } from '../useListCollections';

describe('useListCollections', () => {
	it('should fetch collections successfully', async () => {
		const { result } = renderHook(() => useListCollections());

		// Wait for data to be loaded
		await waitFor(() => {
			expect(result.current.data).toBeDefined();
		});

		// Verify the data matches our mock
		expect(result.current.data).toEqual([mockEthCollection, mockPolCollection]);
		expect(result.current.error).toBeNull();
	});

	it('should handle empty collections', async () => {
		// Mock marketplace config with empty collections
		server.use(
			createLookupMarketplaceConfigHandler({
				...mockConfig,
				collections: [],
			}),
		);

		const { result } = renderHook(() => useListCollections());

		await waitFor(() => {
			expect(result.current.data).toBeDefined();
		});

		expect(result.current.data).toEqual([]);
		expect(result.current.error).toBeNull();
	});

	it('should handle error states', async () => {
		// Mock marketplace config with collection
		server.use(
			createLookupMarketplaceConfigHandler({
				...mockConfig,
				collections: [
					{
						chainId: 1,
						address: '0x1234567890123456789012345678901234567890' as Address,
						feePercentage: 2.5,
						marketplaceType: MarketplaceType.ORDERBOOK,
						currencyOptions: [],
						exchanges: [],
						bannerUrl: '',
						destinationMarketplace: OrderbookKind.sequence_marketplace_v2,
						filterSettings: {
							filterOrder: ['Category', 'Level', 'Element'],
							exclusions: [],
						},
					},
				],
			}),
			http.post('*/rpc/Metadata/GetContractInfoBatch', () => {
				return new HttpResponse(
					JSON.stringify({ error: { message: 'Failed to fetch collections' } }),
					{ status: 500 },
				);
			}),
		);

		const { result } = renderHook(() => useListCollections());

		await waitFor(() => {
			expect(result.current.isError).toBe(true);
		});

		expect(result.current.error).toBeDefined();
		expect(result.current.data).toBeUndefined();
	});

	it('should handle disabled queries', async () => {
		let requestMade = false;

		// Mock marketplace config with collection
		server.use(
			createLookupMarketplaceConfigHandler({
				...mockConfig,
				collections: [
					{
						chainId: 1,
						address:
							'0x1234567890123456789012345678901234567890' as `0x${string}`,
						feePercentage: 2.5,
						marketplaceType: MarketplaceType.ORDERBOOK,
						currencyOptions: [],
						exchanges: [],
						bannerUrl: '',
						destinationMarketplace: OrderbookKind.sequence_marketplace_v2,
						filterSettings: {
							filterOrder: ['Category', 'Level', 'Element'],
							exclusions: [],
						},
					},
				],
			}),
			http.post('*/rpc/Metadata/GetContractInfoBatch', () => {
				requestMade = true;
				return HttpResponse.json({
					contractInfoMap: {
						'0x1234567890123456789012345678901234567890': mockEthCollection,
					},
				});
			}),
		);

		const disabledArgs = {
			query: {
				enabled: false,
			},
		};

		const { result } = renderHook(() => useListCollections(disabledArgs));

		// For disabled queries, we expect no loading state and no data
		expect(result.current.isLoading).toBe(false);
		expect(result.current.data).toBeUndefined();
		expect(result.current.error).toBeNull();
		expect(requestMade).toBe(false);
	});

	it('should handle multiple collections from different chains', async () => {
		// TODO: This test should be more robust, make sure we validate that the marketplace config has multiple chains
		// then the to equal should just check that we are fetching the data from those
		const { result } = renderHook(() => useListCollections());

		await waitFor(() => {
			expect(result.current.data).toBeDefined();
		});

		expect(result.current.data).toEqual([mockEthCollection, mockPolCollection]);
	});
});
