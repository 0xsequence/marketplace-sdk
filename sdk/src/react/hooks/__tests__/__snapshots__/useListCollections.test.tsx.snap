// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useListCollections > should fetch collections successfully 1`] = `
[
  {
    "address": "0x0000000000000000000000000000000000000000",
    "bannerUrl": "",
    "bytecodeHash": "0x1234567890",
    "chainId": 1,
    "currencyOptions": [
      "0x0000000000000000000000000000000000000000",
      "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    ],
    "deployed": true,
    "destinationMarketplace": "sequence_marketplace_v2",
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "feePercentage": 3.5,
    "filterSettings": {
      "exclusions": [
        {
          "condition": "SPECIFIC_VALUE",
          "key": "Type",
          "value": "Sample",
        },
      ],
      "filterOrder": [
        "Type",
        "Rarity",
      ],
    },
    "isLAOSERC721": undefined,
    "itemsAddress": "0x0000000000000000000000000000000000000000",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "MARKET",
    "name": "Mock Collection",
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
  {
    "address": "0x1234567890123456789012345678901234567890",
    "bannerUrl": "https://example.com/collection-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 137,
    "currencyOptions": [
      "0x0000000000000000000000000000000000000000",
    ],
    "deployed": true,
    "destinationMarketplace": "opensea",
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "feePercentage": 2.5,
    "filterSettings": {
      "exclusions": [
        {
          "condition": "ENTIRE_KEY",
          "key": "Category",
        },
        {
          "condition": "SPECIFIC_VALUE",
          "key": "Level",
          "value": "Legendary",
        },
      ],
      "filterOrder": [
        "Category",
        "Level",
        "Element",
      ],
    },
    "isLAOSERC721": undefined,
    "itemsAddress": "0x1234567890123456789012345678901234567890",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "MARKET",
    "name": "Mock Collection",
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
]
`;
